### PR TITLE
Glopes/string constr format

### PR DESCRIPTION
**Pull request by cataphract**
# What Does This Do

Implements taint tracking for String constructors and `String.format`.

### CODE REVIEW

File: internal-api/src/test/groovy/datadog/trace/api/iast/InstrumentationBridgeTest.groovy [hunk 1]
```diff
@@ -1,239 +1,84 @@
...
-    InstrumentationBridge.registerIastModule(module)
-    final self = new StringBuilder()
-
-    when:
-    InstrumentationBridge.onStringBuilderInit(self, 'test')
-
-    then:
-    1 * module.onStringBuilderAppend(self, 'test')
-  }
+  private final static BRIDGE_METHODS = [
```
Comment (manuel-alvarez-alvarez) [inline internal-api/src/test/groovy/datadog/trace/api/iast/InstrumentationBridgeTest.groovy:18]: Nice!

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/IastModuleImpl.java [hunk 2]
```diff
@@ -191,6 +224,214 @@ public void onStringBuilderToString(
...
+      Locale l, String fmt, Object[] args, TaintedObjects taintedObjects, TaintedObject toFmt) {
+    // If the format is tainted, however, things get more complicated.
+    // We find the tainted ranges of the format, replace them with %<n>$s,
+    // and add new arguments at n.
+    // If they include a placeholder, or part of a placeholder, we give up though
+    // (we could attempt to resolve the placeholder first, but at that point
+    // it's not worth the trouble). We mark the whole result as tainted
+
+    // 1st, find all the patterns
+    List<Integer> patternPositions = new ArrayList<>();
```
Comment (jandro996) [inline dd-java-agent/agent-iast/src/main/java/com/datadog/iast/IastModuleImpl.java:269]: Maybe we can avoid the allocation of a new List using a matcher.groupCount approach :)

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/IastModuleImpl.java [hunk 3]
```diff
@@ -191,6 +224,214 @@ public void onStringBuilderToString(
...
+          taintedObjects.taint(
+              result, new Range(0, result.length(), toFmt.getRanges()[0].getSource()));
+          return result;
+        }
+      }
+    }
+
+    // if not, add new placeholders for the ranges, extract the tainted portions,
+    // taint them, and add them as new arguments
+    List<Object> newArgs = new ArrayList<>(Arrays.asList(args));
```
Comment (jandro996) [inline dd-java-agent/agent-iast/src/main/java/com/datadog/iast/IastModuleImpl.java:301]: I think we could avoid the List allocation here using the args length and iterating over the `toFmt.getRanges()` with a counter to calc a newArgs  array size

**Comment by jandro996** [comment]
All the old sources, propagation and sinks PRs must be rebased and refactored due to https://github.com/DataDog/dd-trace-java/pull/4300

**Review by jandro996 — CHANGES_REQUESTED** [review]
All the old sources, propagation and sinks PRs must be rebased and refactored due to https://github.com/DataDog/dd-trace-java/pull/4300

**Pull request closed without merging**